### PR TITLE
fix insert edge

### PR DESCRIPTION
**Pull request by czpmango**
Fixes [https://github.com/vesoft-inc/nebula/issues/2739](https://github.com/vesoft-inc/nebula/issues/2739)

### CODE REVIEW

**Review by yixinglu — COMMENTED** [review]
Fix what problem? could you give some cases ?

**Comment by czpmango** [comment]
> Fix what problem? could you give some cases ?

In the case of multiple replicas, insert statements may cause inconsistent attribute values ​​between replicas.
```
create edge E(prop timestamp default timestamp())
insert edge E() values "1"->"2":()
```

**Comment by Shylock-Hg** [comment]
> > Fix what problem? could you give some cases ?
> 
> In the case of multiple replicas, insert statements may cause inconsistent attribute values ​​between replicas.
> 
> ```
> create edge E(prop timestamp default timestamp())
> insert edge E() values "1"->"2":()
> ```

Between partitions instead of replicas.

**Comment by czpmango** [comment]
> > > Fix what problem? could you give some cases ?
> > 
> > 
> > In the case of multiple replicas, insert statements may cause inconsistent attribute values ​​between replicas.
> > ```
> > create edge E(prop timestamp default timestamp())
> > insert edge E() values "1"->"2":()
> > ```
> 
> Between partitions instead of replicas.

You are right.

**Comment by Shylock-Hg** [comment]
Add case to compare properties of in/out edges

**Comment by Shylock-Hg** [comment]
Close #2867

**Comment by kikimo** [comment]
> Close #2867
@Shylock-Hg 

We are testing TOSS feature, not sure this is the same problem you are confronting.

**Comment by Shylock-Hg** [comment]
> > Close #2867
> > @Shylock-Hg
> 
> We are testing TOSS feature, not sure this is the same problem you are confronting.
> If it's timestamp() mismatch, then #2867 is a whold different problem.

TOSS is another way could resolve this problem.

**Comment by kikimo** [comment]
> > > Close #2867
> > > @Shylock-Hg
> > 
> > 
> > We are testing TOSS feature, not sure this is the same problem you are confronting.
> > If it's timestamp() mismatch, then #2867 is a whold different problem.
> 
> TOSS is another way could resolve this problem.

Sadly TOSS cannot fix too. The problem issue #2867 mentioning is the mismatch of int value of edge props without default. BTW @liuyu85cn has already did some fix for the default timestamp() value mismatch at the storage level.

**Comment by Shylock-Hg** [comment]
> > > > Close #2867
> > > > @Shylock-Hg
> > > 
> > > 
> > > We are testing TOSS feature, not sure this is the same problem you are confronting.
> > > If it's timestamp() mismatch, then #2867 is a whold different problem.
> > 
> > 
> > TOSS is another way could resolve this problem.
> 
> Sadly TOSS cannot fix too. The problem issue #2867 mentioning is the mismatch of int value of edge props without default. BTW @liuyu85cn has already did some fix for the default timestamp() value mismatch at the storage level.

Could resolve the default value issue.

**Comment by czpmango** [comment]
> > Close #2867
> > @Shylock-Hg
> 
> We are testing TOSS feature, not sure this is the same problem you are confronting.
> If it's timestamp() mismatch, then #2867 is a whold different problem.

This pr and issue#2739 has nothing to do with toss, it is just to fix `Default Value` problem.

File: src/graph/validator/MutateValidator.cpp [hunk 1]
```diff
@@ -233,6 +238,32 @@ Status InsertEdgesValidator::prepareEdges() {
...
+      auto iter = std::find(propNames_.begin(), propNames_.end(), propName);
+      if (iter == propNames_.end()) {
+        if (field->hasDefault()) {
+          auto *defaultValue = field->defaultValue();
+          DCHECK(!!defaultValue);
+          auto v = defaultValue->eval(QueryExpressionContext()(nullptr));
+          entirePropValues.emplace_back(v);
+        } else {
+          if (!field->nullable()) {
+            return Status::SemanticError("The not null field doesn't have a default value.");
```
Comment (yixinglu) [inline src/graph/validator/MutateValidator.cpp:255]: refine the error message: 
```
The property `propName' is not nullable and has no default value.
```

File: src/graph/validator/MutateValidator.cpp [hunk 2]
```diff
@@ -233,6 +238,32 @@ Status InsertEdgesValidator::prepareEdges() {
...
+          }
+          entirePropValues.emplace_back(Value(NullType::__NULL__));
+        }
+      } else {
+        auto v = props[std::distance(propNames_.begin(), iter)];
+        if (!field->nullable() && v.isNull()) {
+          return Status::SemanticError("The not null field cannot be null");
```
Comment (yixinglu) [inline src/graph/validator/MutateValidator.cpp:262]: ditto:

```
The non-nullable property `propName' could not be NULL.
```

File: src/graph/validator/MutateValidator.cpp [hunk 3]
```diff
@@ -233,6 +238,32 @@ Status InsertEdgesValidator::prepareEdges() {
     auto valsRet = SchemaUtil::toValueVec(row->values());
     NG_RETURN_IF_ERROR(valsRet);
     auto props = std::move(valsRet).value();
+
+    std::vector<Value> entirePropValues;
+    for (size_t j = 0; j < fieldNum; ++j) {
+      auto *field = schema_->field(j);
+      auto propName = entirePropNames_[j];
```
Comment (yixinglu) [inline src/graph/validator/MutateValidator.cpp:245]: It seems that the `entirePropNames_` is not necessary, why not use `schema_->field(j)->name()` here directly. And you needs not to save it as the field of `MutateValidator`

File: src/graph/validator/MutateValidator.cpp (hunk 2 above)
Comment (czpmango) [reply src/graph/validator/MutateValidator.cpp:262]: ok. I copied previous error messege in storage..

File: src/graph/validator/MutateValidator.cpp (hunk 3 above)
Comment (czpmango) [reply src/graph/validator/MutateValidator.cpp:245]: ACK.

File: src/graph/validator/MutateValidator.cpp (hunk 1 above)
Comment (czpmango) [reply src/graph/validator/MutateValidator.cpp:255]: ACK.

File: src/graph/validator/MutateValidator.cpp (hunk 3 above)
Comment (czpmango) [reply src/graph/validator/MutateValidator.cpp:245]: Also used in `InsertEdgesValidator::toPlan()`.

**Review by yixinglu — APPROVED** [review]

**Review by Aiee — APPROVED** [review]

**Pull request merged**